### ISSUE TITLE
Add ability to bookmark or flag individual lines of code

### ISSUE THREAD

**Issue by jafflerbach** [rstudio/rstudio#4657]
@marinaesp and I are submitting a feature request to add the ability to symbolically flag, bookmark or otherwise identify a line or lines of code in an R/Rmd script on the left hand side of the script pane. 

Sometimes I find myself scrolling through a script trying to remember where I was working an hour ago or a week ago. Or I see a simple bug or typo in a script that I don't have time to work on at the moment but want to make sure I come back to later on.

Identifying these areas for future reference is rather difficult right now. Some suggestions include using `====` or `####` but that does not seem to work within Rmarkdown code chunks. It has also been suggested to use code sections with Command + Shift + R to create a section header, but again it seems this is meant for use in .R scripts. I almost exclusively use .Rmd and would love the ability to hover over a code line number on the left hand side, right click, and add a "flag" or "bookmark" that would help me go back to that location, either later that day or later that week. A lot of R users have their own, bespoke ways of doing this whether it is using the solutions mentioned above or writing in all caps after multiple #, such as:

```r
#### NEED TO FIGURE OUT WHY COLUMN X IS BEING DROPPED BELOW. START HERE ON MONDAY!!!!

data <- raw %>%
   left_join(other_table)
```

Taking this a step further, the flag could allow a text box to pop up for us to add a comment (not the traditional `#comment`). This would be similar to comments in Microsoft word or Adobe acrobat. Here is my PowerPoint rendering of how this *could* look.

![image](https://user-images.githubusercontent.com/6842510/56306602-bf0d0180-60f7-11e9-8652-024549b65eb6.png)

Based on [this twitter response](https://twitter.com/jafflerbach/status/1118529591948136448), I do think there would be greater interest from the R community for this feature.

**Comment by jmcphers** [comment]
Thanks for this idea!

One thing you might try playing with is `rstudioapi::sourceMarkers` (see [documentation](https://rdrr.io/cran/rstudioapi/man/sourceMarkers.html)), which will allow you to create a list of source annotations. These are bookmarks with any text you like, and they show up in a Markers pane so it's easy to browse your list of notes and jump back. 

![image](https://user-images.githubusercontent.com/470418/56309315-304fb300-60fe-11e9-904d-e478f2a80175.png)

You could pretty simply build an [RStudio add-in](https://rstudio.github.io/rstudioaddins/) which prompts the user for text and then adds a marker at the given line. I don't think it's an exact match for what you want, but might be helpful!

**Comment by marinaesp** [comment]
Hello Jonathan, 
Thank you for responding to our request.
I tried toe sourceMarker, but did not really get it to work. Would you be able to give a simple example how you invoke a marker with a line of code?
And thanks for the tip about add-ins, I will check them out!

**Comment by jmcphers** [comment]
I made an add-in (just for fun, it's not very polished!). You can try it here:

https://github.com/jmcphers/bookmarkr

**Comment by marinaesp** [comment]
Thank you for efforts, Jonathan. 
**BookmarkR** works! 
Please, let me know if you also come up with a way to delete a bookmark or add a coloured symbol to a bookmark line.

Many thanks!!

Marina

**Comment by earcanal** [comment]
As a bookmark hack, I'm using the NOTE marker with

  https://github.com/dokato/todor

It's also breathed new life into all of my FIXME and TODO comments :)

**Comment by jbullock35** [comment]
The todor add-on that @earcanal mentions is useful. But what one would really like, I think, is what other IDEs have: the ability to toggle a "bookmark" symbol in the left-hand margin, next to the line numbers, by issuing a keyboard shortcut. Coupled with the ability to jump to a bookmark, or cycle through bookmarks, with another keyboard shortcut. 

Being able to refer to bookmarks only by looking to the "Markers" pane is not a good substitute. One should be able to see bookmarks in the left-hand margin of one's document.

**Comment by GitHunter0** [comment]
@jbullock35 , exactly, that would be the most useful. Stata has that feature, might serve as a reference.

**Comment by Nova-Scotia** [comment]
> Thank you for efforts, Jonathan. **BookmarkR** works! Please, let me know if you also come up with a way to delete a bookmark or add a coloured symbol to a bookmark line.
> 
> Many thanks!!
> 
> Marina

I'm using bookmarkr, it's great!
